ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java

```diff
@@ -0,0 +1,842 @@
+        getIndexEntriesToCopy(allIndexEntries, duplicateSearchSpan, indexSegmentToCopy.getStartOffset(),
+            cutoffOffsetForDelete, checkAlreadyCopied);
+
+    // Copy these over
+    boolean copiedAll = copyRecords(from, indexEntriesToCopy, indexSegmentToCopy.getLastModifiedTimeSecs());
+    // persist
+    tgtIndex.persistIndex();
+    return copiedAll;
+  }
+
+  /**
+   * Gets all the index entries that need to be copied. The index entries that will be copied are:
+   * 1. Records that are not duplicates.
+   * 2. Records that are not expired or deleted.
+   * 3. Delete records.
+   * 4. Records that are already in the target log (because they were copied before crash/shutdown).
+   * @param allIndexEntries the {@link MessageInfo} of all the entries in the index segment.
+   * @param duplicateSearchSpan the {@link FileSpan} in which to search for duplicates
+   * @param indexSegmentStartOffset the start {@link Offset} of the index segment under consideration.
+   * @param cutoffOffsetForDelete the {@link Offset} up until which delete records are considered applicable.
+   * @param checkAlreadyCopied {@code true} if a check for existence in the swap spaces has to be executed (due to
+   *                                       crash/shutdown), {@code false} otherwise.
+   * @return the list of valid {@link IndexEntry} sorted by their offset.
+   * @throws IllegalStateException if there is a mismatch b/w the index value expected and the index value obtained.
+   * @throws StoreException if there is any problem with using the index.
+   */
+  private List<IndexEntry> getIndexEntriesToCopy(List<IndexEntry> allIndexEntries, FileSpan duplicateSearchSpan,
+      Offset indexSegmentStartOffset, Offset cutoffOffsetForDelete, boolean checkAlreadyCopied) throws StoreException {
+    List<IndexEntry> indexEntriesToCopy = new ArrayList<>();
+    List<IndexEntry> copyCandidates =
+        getValidIndexEntries(indexSegmentStartOffset, allIndexEntries, cutoffOffsetForDelete);
+    for (IndexEntry copyCandidate : copyCandidates) {
+      IndexValue copyCandidateValue = copyCandidate.getValue();
+      // search for duplicates in srcIndex if required
+      if (duplicateSearchSpan != null) {
+        IndexValue possibleDuplicate = srcIndex.findKey(copyCandidate.getKey(), duplicateSearchSpan);
+        if (possibleDuplicate != null) {
+          if (possibleDuplicate.isFlagSet(IndexValue.Flags.Delete_Index)) {
+            // copyCandidate is surely a duplicate because srcIndex contains a DELETE index entry
+            continue;
+          } else if (!copyCandidateValue.isFlagSet(IndexValue.Flags.Delete_Index)) {
+            // copyCandidate is a duplicate because it is a PUT entry and so is the possible duplicate.
+            continue;
+          }
+        }
+      }
+      // search for duplicates in tgtIndex if required
+      if (!checkAlreadyCopied || !alreadyExistsInTgt(copyCandidate.getKey(), copyCandidateValue)) {
```

Comment:
It won't result in a look up for every candidate. It will result in a lookup when `checkAlreadyCopied` is `true` which happens only on recovery and for exactly one index segment (the first one to be copied after recovery).